{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project Description\n",
    "### *Trypanosoma brucei* performs RNA editing on several mitochondrially expressed gene transcripts. This editing takes place in the form of uridine insertions or deletions. In a population of transcripts generated from a single gene, many different forms can exist, with different amounts of Us at different places. The goal of this project is to align multiple sequences of the same gene transcript at different stages of editing.\n",
    "\n",
    "### This tool is applicable to sequences of edited mRNAs generated in all kinetoplastid protozoa.  It is not designed for other types of RNA editing, such as A to I, but could possibly work those, as the basis of this program is a sequence alignment program.  However, this would be quite unnecessary as the special treatment of Us in this program is a waste of time for an A to I edited transcript.  \n",
    "\n",
    "### So far, I have only tested this program on 31 total sequences.  I will be testing it soon on larger data sets, but probably no larger than 100 sequences, as I do not have this data currently.  I have tested it on two genes so far, and plan to test it on more data, as I assemble it.  There are only 12 edited genes in *T. brucei*."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence Class\n",
    "### To perform this task, we need to generate several pieces of information from each sequence, such as the name of the sequence, the sequence itself, the sequence with Us removed and the list of the number of Us at each potential U site.  To tie these pieces together, I made Sequence class.  This class is only able to extract the information if it is given in fasta format.  It also converts the sequences to uppercase, removes asterisks and Ts.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Sequence:\n",
    "    def __init__(self,string):\n",
    "        self.content=string\n",
    "        #remove *s, us, and Ts\n",
    "        self.content=self.content.replace(\"*\",\"\")\n",
    "        self.content=self.content.replace(\"u\",\"U\")\n",
    "        self.content=self.content.replace(\"T\",\"U\")\n",
    "        self.content=self.content.replace(\"a\",\"A\")\n",
    "        self.content=self.content.replace(\"g\",\"G\")\n",
    "        self.content=self.content.replace(\"c\",\"C\")\n",
    "\n",
    "        #count the length of the name\n",
    "        self.titleLength=0\n",
    "        while self.content[self.titleLength]!=(\"\\n\"):\n",
    "            self.titleLength=self.titleLength+1\n",
    "        \n",
    "        #make the title\n",
    "        self.title=\"\"\n",
    "        for i in range(self.titleLength-1):\n",
    "            self.title=self.title+self.content[i+1]#this excludes the \">\" at the beginning of each line\n",
    "        \n",
    "        #extract the sequence by looking after the title\n",
    "        self.sequence=self.content[self.titleLength+1]\n",
    "        for i in range(self.titleLength+2,len(self.content)):\n",
    "            self.sequence=self.sequence+self.content[i]\n",
    "        self.sequence=self.sequence.replace(\"\\n\",\"\")#remove linebreaks\n",
    "        \n",
    "        #make the noU sequence and the U list\n",
    "        self.uList=list()\n",
    "        uCounter=0 #counts the number of Us at any given position\n",
    "        seqPos=0 #is the placekeeper in the sequence\n",
    "        self.noUs=\"\"\n",
    "        #initialize the uList\n",
    "        if self.sequence[0]==\"U\":\n",
    "            uCounter=1\n",
    "            seqPos=1\n",
    "            #count how many consecutive Us are at this position\n",
    "            while self.sequence[uCounter]==\"U\":\n",
    "                uCounter=uCounter+1\n",
    "                seqPos=seqPos+1\n",
    "            self.uList.append(uCounter)\n",
    "        else:\n",
    "            self.uList.append(0)\n",
    "        #add each non-U nucleotide to the noUs string and count how many Us come after each nucleotide\n",
    "        while seqPos<len(self.sequence):\n",
    "            self.noUs=self.noUs+self.sequence[seqPos]\n",
    "            seqPos=seqPos+1\n",
    "            uCounter=0\n",
    "            if seqPos<len(self.sequence):\n",
    "                if self.sequence[seqPos]==\"U\":\n",
    "                    uCounter=1\n",
    "                    seqPos=seqPos+1\n",
    "                    while seqPos<len(self.sequence) and self.sequence[seqPos]==\"U\":\n",
    "                        uCounter=uCounter+1\n",
    "                        seqPos=seqPos+1\n",
    "                    self.uList.append(uCounter)\n",
    "                else:\n",
    "                    self.uList.append(0)\n",
    "        #for some reason, it wasn't adding the U at the end every time, so I added these two lines\n",
    "        if self.sequence[seqPos-1]!=\"U\":\n",
    "            self.uList.append(0)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# extractSequences function\n",
    "### This function serves to open a text file with a list of sequences in the fasta format and extract each sequence.  It saves all of these in a list and saves each sequence as the class, Sequence (see above), in a new list, allSeq, which is returned.  \n",
    "### In the fasta format, each sequence is denoted by a \">\" followed by the name of the sequence, with the actual sequence on the following line.  This function will fail if sequences are not input in this method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def extractSequences(filename):\n",
    "    F=open(filename)\n",
    "    content=F.readlines()\n",
    "    F.close()\n",
    "    sequences=list() #dummy list to hold raw sequence information\n",
    "    for lines in range(len(content)):\n",
    "        #find the start of the sequence\n",
    "        if content[lines][0]==\">\":\n",
    "            sequence=\"\"\n",
    "            sequence=sequence+content[lines]+content[lines+1]\n",
    "            sequences.append(sequence)\n",
    "    allSeq=list()\n",
    "    #convert all raw sequences to the Sequence class in a new list\n",
    "    for i in range(len(sequences)):\n",
    "        seqi=Sequence(sequences[i])\n",
    "        allSeq.append(seqi)\n",
    "    return allSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 10 sequences in this file.\n"
     ]
    }
   ],
   "source": [
    "ND3shortSeq=extractSequences(\"ND3align.fa\")\n",
    "print \"There are {0} sequences in this file.\".format(len(ND3shortSeq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scoreMatrix function\n",
    "### This function serves to generate a score matrix for two sequences.  It does this by creating a table that is one column longer than the first sequence and one row longer than than the second sequence.  This first row and column are filled with zeros, the second row at the second position, is then scored by evaluating the first nucleotide pair.  Scores for all other cells are filled in using the scores of cells left, above and above and left to determine which will yield the highest value of the three.  If the nucleotides match, the above and left cell score recieves a matching bonus, but if they do not it receives a mismatch penalty.  Above scores and left scores receive a gap penalty.  \n",
    "\n",
    "### The gap penalty, mismatch penalty and match bonus were not chosen at random.  These were selected based on which set of numbers generated good outputs.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def scoreMatrix(seq1,seq2):\n",
    "    #initialize the matrix\n",
    "    matrix=list()\n",
    "    for i in range(len(seq1)+1):\n",
    "        matrix.append(list())\n",
    "        for j in range(len(seq2)+1):\n",
    "            matrix[i].append(0.0)\n",
    "\n",
    "    #fill it in with scores\n",
    "    k=1 #keep track of position left to right\n",
    "    l=1 #keep track of position up and down\n",
    "    gapPenalty=-1\n",
    "    mismatchPenalty=0\n",
    "    matchScore=1\n",
    "    while l<(len(seq2)+1):\n",
    "        while k<(len(seq1)+1):\n",
    "            score1=matrix[k][l-1]+gapPenalty#above\n",
    "            score2=matrix[k-1][l]+gapPenalty#left\n",
    "            score3=matrix[k-1][l-1]+mismatchPenalty#above and to the left\n",
    "            if seq1[k-1]==seq2[l-1]:\n",
    "                score3=score3+(matchScore-mismatchPenalty)\n",
    "            matrix[k][l]=max(score1,score2,score3)\n",
    "            k=k+1\n",
    "        l=l+1\n",
    "        k=1\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n",
      "181\n",
      "49\n",
      "50\n"
     ]
    }
   ],
   "source": [
    "ND30and1matrix=scoreMatrix(ND3shortSeq[0].noUs,ND3shortSeq[1].noUs)\n",
    "print len(ND3shortSeq[0].noUs)\n",
    "print len(ND30and1matrix) #should be one greater\n",
    "print len(ND3shortSeq[1].noUs)\n",
    "print len(ND30and1matrix[0]) #should be one greater\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dashPositions function and makeAlignment function\n",
    "### The dashPositions function uses the matrix previously generated and the sequences used to generate it to create two lists that will later dictate where dashes should be inserted to create the sequence alignment in makeAlignment.  This occurs by first identifiying the highest end score and tracing back through the matrix along the path with the highest scores.  The two lists record the position along each sequence as the function steps through the score matrix.  \n",
    "\n",
    "### The makeAlignment function looks at the two lists from dashPositions and generates two aligned sequences.  It does this by reading each position in the list and its next neighbor.  If the two positions have the same number, a dash is appended to the new aligned sequence, if not, the nucleotide in the sequence is added.   This continues to the end of the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def dashPositions(seq1,seq2,matrix):\n",
    "    #Find the position with the highest final score, which is the end of the best alignment\n",
    "    maxScore=-1000\n",
    "    m=len(seq1)#position left to right\n",
    "    n=len(seq2)#position up and down\n",
    "    \n",
    "    #This if statement makes it start searching for the highest score along the length of the longer of the two sequences\n",
    "    if len(seq1)>len(seq2):\n",
    "        for i in range(len(matrix)):\n",
    "            if matrix[i][(len(matrix[i])-1)]>maxScore:\n",
    "                maxScore=matrix[i][(len(matrix[i])-1)]\n",
    "                m=i\n",
    "                n=(len(matrix[i]))-1\n",
    "        for i in range(len(matrix[0])):\n",
    "            if matrix[len(matrix)-1][i]>maxScore:\n",
    "                m=(len(matrix))-1\n",
    "                n=i\n",
    "    else:\n",
    "        for i in range(len(matrix[0])):\n",
    "            if matrix[len(matrix)-1][i]>maxScore:\n",
    "                m=(len(matrix))-1\n",
    "                n=i\n",
    "        for i in range(len(matrix)):\n",
    "            if matrix[i][(len(matrix[i])-1)]>maxScore:\n",
    "                maxScore=matrix[i][(len(matrix[i])-1)]\n",
    "                m=i\n",
    "                n=(len(matrix[i]))-1\n",
    "        \n",
    "    \n",
    "    Ms=list()\n",
    "    Ns=list()\n",
    "    alignmentScore=matrix[m][n]\n",
    "    left=0\n",
    "    up=0\n",
    "    \n",
    "    #count up all the ms and ns\n",
    "    while (m+n)!=0:\n",
    "        Ms.append(m)\n",
    "        Ns.append(n)\n",
    "        if m==0:\n",
    "            n=n-1\n",
    "        else:\n",
    "            if n==0:\n",
    "                m=m-1\n",
    "            else:            \n",
    "                score1=matrix[m][n-1]#above\n",
    "                score2=matrix[m-1][n]#left\n",
    "                score3=matrix[m-1][n-1]#above and to the left\n",
    "                #the up and left counters prevent \"stair stepping\" which generates alignments that look like:\n",
    "                #G-A-C-G-A\n",
    "                #-G-A-C-G-A\n",
    "                if score1>score2 and score1>score3 and left==0:\n",
    "                    n=n-1\n",
    "                    up=1\n",
    "                    left=0\n",
    "                else:\n",
    "                    if score2>score3 and up==0:\n",
    "                        m=m-1\n",
    "                        left=1\n",
    "                        up=0\n",
    "                    else:\n",
    "                        m=m-1\n",
    "                        n=n-1\n",
    "                        left=0\n",
    "                        up=0\n",
    "               \n",
    "    Ms.reverse()\n",
    "    Ns.reverse()\n",
    "    \n",
    "    #add in any missing numbers from not starting at the very ends of the sequences\n",
    "    \n",
    "    #these add in numbers that account for the nucleotides after the end of the best alignment\n",
    "    while max(Ms)<len(seq1):\n",
    "        Ms.append(max(Ms)+1)\n",
    "    while max(Ns)<len(seq2):\n",
    "        Ns.append(max(Ns)+1)\n",
    "    #these add in numbers that generate dashes that make the two sequences the same length\n",
    "    while len(Ms)<len(Ns):\n",
    "        Ms.append(max(Ms))\n",
    "    while len(Ns)<len(Ms):\n",
    "        Ns.append(max(Ns))\n",
    "\n",
    "    \n",
    "    return Ms,Ns,alignmentScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n",
      "180\n",
      "48.0\n"
     ]
    }
   ],
   "source": [
    "ND3seq0Pos,ND3seq1Pos,ND30and1Score=dashPositions(ND3shortSeq[0].noUs,ND3shortSeq[1].noUs,ND30and1matrix)\n",
    "print len(ND3seq0Pos)\n",
    "print len(ND3seq1Pos)#should be equal\n",
    "print ND30and1Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def makeAlignment(seq1,seq2,Ms,Ns):\n",
    "    \n",
    "    #start seq1align with either dashes or the first nucleotide\n",
    "    y=0\n",
    "    seq1align=\"\"\n",
    "    while Ms[y]==0:\n",
    "        seq1align=seq1align+\"-\"\n",
    "        y=y+1\n",
    "    seq1align=seq1align+seq1[0]\n",
    "\n",
    "    #start seq2align with either dashes or the first nucleotide\n",
    "    z=0\n",
    "    seq2align=\"\"\n",
    "    \n",
    "    while Ns[z]==0:\n",
    "        seq2align=seq2align+\"-\"\n",
    "        z=z+1\n",
    "    seq2align=seq2align+seq2[0]    \n",
    "    \n",
    "    #Fill in seq1align\n",
    "    o=0 #seq1 position counter \n",
    "    for i in range(y,len(Ms)-1):\n",
    "        if Ms[i]==Ms[i+1]:\n",
    "            seq1align=seq1align+\"-\"\n",
    "        else:\n",
    "            seq1align=seq1align+seq1[o+1]\n",
    "            o=o+1\n",
    "    #Fill in seq2align\n",
    "    p=0 #seq2 position counter \n",
    "    for i in range(z,len(Ns)-1):\n",
    "        if Ns[i]==Ns[i+1]:\n",
    "            seq2align=seq2align+\"-\"\n",
    "        else:\n",
    "            seq2align=seq2align+seq2[p+1]\n",
    "            p=p+1\n",
    "\n",
    "\n",
    "    return seq1align,seq2align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n",
      "180\n",
      "CAAAAAACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGAGCAGAGCGGGGGGCGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGGGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAACAAAAAAAAAAAAAAA\n",
      "-------------------------------------------------------------------------------------------------------------------GGGGGGGG-AGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n"
     ]
    }
   ],
   "source": [
    "ND3align0,ND3align1=makeAlignment(ND3shortSeq[0].noUs,ND3shortSeq[1].noUs,ND3seq0Pos,ND3seq1Pos)\n",
    "print len(ND3align0)\n",
    "print len(ND3align1)#should be equal\n",
    "print ND3align0\n",
    "print ND3align1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# quickAlign function\n",
    "### This function ties together the scoreMatrix, dashPositions, and makeAlignment functions, so you can input two sequences and return their alignment and score (or their matrix and dash position lists, for diagnostic purposes). This function does not return a perfect alignment as it does not correct for stumps at the 5' end.\n",
    "### Example:\n",
    "### Correct alignment:\n",
    "### GGGGGGGGGAGA\n",
    "### -GGGGGGGGAGA\n",
    "### What is actually returned:\n",
    "### GGGGGGGGGAGA\n",
    "### GGGGGGGG-AGA\n",
    "### This is wrong, because later, editing positions at these ends will be misaligned.  However, this is corrected in a later function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def quickAlign(seq1,seq2):\n",
    "    matrix=scoreMatrix(seq1,seq2)\n",
    "    Ms,Ns,alignmentScore=dashPositions(seq1,seq2,matrix)\n",
    "    seq1align,seq2align=makeAlignment(seq1,seq2,Ms,Ns)\n",
    "    return seq1align,seq2align,alignmentScore,#matrix,Ms,Ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAAAAAACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGAGCAGAGCGGGGGGCGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGGGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAACAAAAAAAAAAAAAAA\n",
      "-------------------------------------------------------------------------------------------------------------------GGGGGGGG-AGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n",
      "48.0\n"
     ]
    }
   ],
   "source": [
    "ND3quick0,ND3quick1,ND3quickScore=quickAlign(ND3shortSeq[0].noUs,ND3shortSeq[1].noUs)\n",
    "print ND3quick0\n",
    "print ND3quick1\n",
    "print ND3quickScore"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findBest function\n",
    "### The findBest function looks at a group of sequences, finds the best aligning pair and aligns them.  Then it stores their index numbers in a second list called bestList.  If this list does not exist, it creates a new one, and if it does exist it adds new numbers to the existing list.  \n",
    "### Also, if a bestList does exist, it will search for the best pair that includes one sequence on the bestList and one sequence not on the bestList.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def findBest(seqList,bestList):\n",
    "    bestScore=-1000\n",
    "    if bestList==None:\n",
    "        bestList=list()\n",
    "        #align all sequences\n",
    "        for i in range(len(seqList)):\n",
    "            for j in range(len(seqList)):\n",
    "                if i!=j:\n",
    "                    first,second,score=quickAlign(seqList[i].noUs,seqList[j].noUs)\n",
    "                    #if this alignment is better than the current best, record this as the best\n",
    "                    if score>bestScore:\n",
    "                        bestScore=score\n",
    "                        index1=i \n",
    "                        index2=j\n",
    "                        best1=first\n",
    "                        best2=second\n",
    "        bestList.append(index1)\n",
    "        bestList.append(index2)\n",
    "    \n",
    "    else:\n",
    "        #look at only the sequences in the best list\n",
    "        for i in bestList:\n",
    "            #and compare them to the sequences not in the best list\n",
    "            for j in range(len(seqList)):\n",
    "                if j not in bestList:\n",
    "                    first,second,score=quickAlign(seqList[i].noUs,seqList[j].noUs)\n",
    "                    if score>bestScore:\n",
    "                        bestScore=score\n",
    "                        index1=i\n",
    "                        index2=j\n",
    "                        best1=first\n",
    "                        best2=second\n",
    "        #add the new guy to the best list\n",
    "        bestList.append(index2)\n",
    "        seqList[index2].noUs=best2\n",
    "        \n",
    "        #fix all the old best list members to match the new one\n",
    "        #see how the best1 is different\n",
    "        differences=list() #positions of the differences\n",
    "        whatDifferences=list() #what the differences actually are\n",
    "        n=0 #counter of the adjusted best1\n",
    "        i=0 #counter of the original best1\n",
    "        while i!=(len(seqList[index1].noUs)-1):\n",
    "            if best1[n]!=seqList[index1].noUs[i]:\n",
    "                differences.append(n)\n",
    "                whatDifferences.append(best1[n])\n",
    "                n=n+1\n",
    "            else:\n",
    "                n=n+1\n",
    "                i=i+1\n",
    "        numOfDif=len(differences)\n",
    "        #this corrects for dashes added to the end of best1\n",
    "        if (len(best1)-numOfDif)>len(seqList[index1].noUs):\n",
    "            for i in range(len(best1)-numOfDif-len(seqList[index1].noUs)):\n",
    "                differences.append(i+len(seqList[index1].noUs)+numOfDif)\n",
    "                whatDifferences.append(best1[i+len(seqList[index1].noUs)+numOfDif])\n",
    "        #update best1\n",
    "        seqList[index1].noUs=best1\n",
    "        #update all others\n",
    "        if len(differences)>0:\n",
    "            for j in bestList:\n",
    "                if j!=index1 and j!=index2:\n",
    "                    dummyseq=seqList[j].noUs\n",
    "                    for k in range(len(differences)):\n",
    "                        if differences[k]<len(seqList[j].noUs):\n",
    "                            dummyseq=dummyseq[:differences[k]]+whatDifferences[k]+dummyseq[differences[k]:]\n",
    "                        else:\n",
    "                            dummyseq=dummyseq+whatDifferences[k]\n",
    "                    seqList[j].noUs=dummyseq\n",
    "                    \n",
    "    \n",
    "    #add all updated sequences to new list\n",
    "    newSeqList=list()\n",
    "    for i in range(len(seqList)):\n",
    "        if (i==index1):\n",
    "            seqList[i].noUs=best1\n",
    "        else:\n",
    "            if i==index2:\n",
    "                seqList[i].noUs=best2\n",
    "        newSeqList.append(seqList[i])    \n",
    "    \n",
    "    \n",
    "    return newSeqList,bestList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 3]\n"
     ]
    }
   ],
   "source": [
    "ND3shortNew,ND3best=findBest(ND3shortSeq,None)\n",
    "print ND3best #should have two items"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# stumpRepair function\n",
    "### This function repairs the stump issue I previously mentioned in the quickAlign function description.  It checks to see if any sequence separated by a gap from the rest of the sequence could be aligned without the gap.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def stumpRepair(seq1align,seq2align):\n",
    "    gapCounter1=0 #counts how long a gap is between two pieces of sequence\n",
    "    stumpCounter1=0 #counts how long a piece of sequence is before a gap\n",
    "    breakPoint1=0 #records where in the sequence the gap ends\n",
    "    gapCounter2=0\n",
    "    stumpCounter2=0\n",
    "    breakPoint2=0\n",
    "    #are there any stumps?\n",
    "    for i in range(len(seq1align)-1):\n",
    "        #if there is a letter and we haven't counted gaps yet, count the stump\n",
    "        if seq1align[i]!=\"-\" and gapCounter1==0:\n",
    "            stumpCounter1=stumpCounter1+1\n",
    "        #if there is a dash and we have a counted stump, count the gap\n",
    "        if seq1align[i]==\"-\" and stumpCounter1!=0:\n",
    "            gapCounter1=gapCounter1+1\n",
    "            #when we hit the next letter stop counting and record the break point\n",
    "            if seq1align[i+1]!=\"-\":\n",
    "                breakPoint1=i\n",
    "                break\n",
    "        #repeat above for second sequence\n",
    "        if seq2align[i]!=\"-\" and gapCounter2==0:\n",
    "            stumpCounter2=stumpCounter2+1\n",
    "        if seq2align[i]==\"-\" and stumpCounter2!=0:\n",
    "            gapCounter2=gapCounter2+1\n",
    "            if seq2align[i+1]!=\"-\":\n",
    "                breakPoint2=i\n",
    "                break\n",
    "    \n",
    "    #can we shift the stump?\n",
    "    if breakPoint1!=0:\n",
    "        canShift=0\n",
    "        for j in range(stumpCounter1):\n",
    "            #if the each nt of the stump matches where they would line up without the gap, canShift+1\n",
    "            if seq1align[breakPoint1-gapCounter1-stumpCounter1+j+1]==seq2align[breakPoint1-stumpCounter1+j+1]:\n",
    "                canShift=canShift+1\n",
    "        #if they all match, then canShift will equal the length of the stump\n",
    "        if canShift==stumpCounter1:\n",
    "            dummy=\"\"\n",
    "            #add in alignment before the stump\n",
    "            for k in range(breakPoint1-stumpCounter1-gapCounter1+1):\n",
    "                dummy=dummy+seq1align[k]\n",
    "            #add in appropriate number of dashes\n",
    "            for l in range(gapCounter1):\n",
    "                dummy=dummy+\"-\"\n",
    "            #add in stump\n",
    "            for m in range(stumpCounter1):\n",
    "                dummy=dummy+seq1align[breakPoint1-stumpCounter1-gapCounter1+m+1]\n",
    "            #add in the rest of the sequence\n",
    "            for n in range(len(seq1align)-breakPoint1-1):\n",
    "                dummy=dummy+seq1align[breakPoint1+n+1]\n",
    "            seq1align=dummy\n",
    "    if breakPoint2!=0:\n",
    "        canShift=0\n",
    "        for j in range(stumpCounter2):\n",
    "            if seq2align[breakPoint2-gapCounter2-stumpCounter2+j+1]==seq1align[breakPoint2-stumpCounter2+j+1]:\n",
    "                canShift=canShift+1\n",
    "        if canShift==stumpCounter2:\n",
    "            dummy=\"\"\n",
    "            for k in range(breakPoint2-stumpCounter2-gapCounter2+1):\n",
    "                dummy=dummy+seq2align[k]\n",
    "            for l in range(gapCounter2):\n",
    "                dummy=dummy+\"-\"\n",
    "            for m in range(stumpCounter2):\n",
    "                dummy=dummy+seq2align[breakPoint2-stumpCounter2-gapCounter2+m+1]\n",
    "            for n in range(len(seq2align)-breakPoint2-1):\n",
    "                dummy=dummy+seq2align[breakPoint2+n+1]\n",
    "            seq2align=dummy\n",
    "\n",
    "    return seq1align,seq2align\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAAAAAACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGAGCAGAGCGGGGGGCGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGGGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAACAAAAAAAAAAAAAAA\n",
      "-------------------------------------------------------------------------------------------------------------------GGGGGGGG-AGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n",
      "CAAAAAACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGAGCAGAGCGGGGGGCGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGGGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAACAAAAAAAAAAAAAAA\n",
      "--------------------------------------------------------------------------------------------------------------------GGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n"
     ]
    }
   ],
   "source": [
    "print ND3quick0\n",
    "print ND3quick1\n",
    "ND3New0,ND3New1=stumpRepair(ND3quick0,ND3quick1)\n",
    "print ND3New0\n",
    "print ND3New1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# alignAll function\n",
    "### Like quickAlign, alignAll ties together multiple functions, in this case, findBest and stump repair.  It repeats the use of findBest until all of the sequences have been incorporated to the best list, and therefore aligned.  Then it performs the stumpRepair function on all pairs of sequences to ensure that no stumps are misaligned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def alignAll(seqList):\n",
    "    newSeqList=list()\n",
    "    #find the first best pair\n",
    "    newSeqList,bestList=findBest(seqList,None)\n",
    "\n",
    "    #match all sequences to the best sequences\n",
    "    while len(newSeqList)!=len(bestList):\n",
    "        newSeqList,bestList=findBest(newSeqList,bestList)\n",
    "     \n",
    "    #repair stumps by checking all pairs\n",
    "    fixedList=list()\n",
    "    for j in range(len(newSeqList)):\n",
    "        newj=newSeqList[j].noUs\n",
    "        for k in range(len(newSeqList)):\n",
    "            if j!=k:\n",
    "                newk=\"\"\n",
    "                newj,newk=stumpRepair(newj,newSeqList[k].noUs)\n",
    "        fixedList.append(newj)\n",
    "        seqList[j].noUs=newj\n",
    "\n",
    "    return seqList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ND3Conv\n",
      "CAAAAAACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGAGCAGAGCGGGGGGCGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGGGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAACAAAAAAAAAAAAAAA\n",
      "ND3AltF01\n",
      "--------------------------------------------------------------------------------------------------------------------GGGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n",
      "ND3AltA02\n",
      "-----AACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGG--------------------------------------------------------\n",
      "ND3Alt4E01\n",
      "-----AACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGAGGGGAGGGACACG------------------------------------\n",
      "ND3Alt2A03\n",
      "---------------------------------------------------------------------------------------------------------------------GGGGGGGAGAAGGAGGGGAGGGACACGAAGGGGGAAAGAGAGGGAAAA---------------\n",
      "ND3Alt2C02\n",
      "-----AACCCGCCACAGGACAAGAGGAAGGGAAAAAGGGGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGA-------------------------------------------------\n",
      "ND3AltD02\n",
      "-----AACCCGCCACAGGACAAAGGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGA-------------------------------------------------\n",
      "ND3AltE02\n",
      "-----AACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGGGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGA-------------------------------------------------\n",
      "ND3Alt3B01\n",
      "-----AACCCGCCACAGGACAAAGGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACAGGGG-AACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGA-------------------------------------------------\n",
      "ND3Alt3B02\n",
      "-----AACCCGCCACAGGACAAAAGGAAGGGAAAAAGGAGAGGGCAGAGCGGGGGACGGGGGAGAAGAACACGGGGAACAGGAAGGAGGGGAGAAACCAAGGGAGGGAAGGGAGAGGGGGGGGGAGAAGGA-------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "ND3alignedShort=alignAll(ND3shortSeq)\n",
    "for i in range(len(ND3alignedShort)):\n",
    "    print ND3alignedShort[i].title\n",
    "    print ND3alignedShort[i].noUs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# updateUs function\n",
    "### This function looks at an aligned sequence and identifies the locations of dashes.  Because dashes indicate that there are other potential editing sites, it then goes to the corresponding position in the u list and adds a zero there.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def updateUs(ulist,alignedSeq):\n",
    "\n",
    "    newUlist=list() \n",
    "    oldUcounter=0 #keeps position in the old U list\n",
    "    alignCounter=0 #keeps position in the aligned sequence\n",
    "    while alignCounter<len(alignedSeq):\n",
    "        #if there is a dash, put in a zero at the corresponding position\n",
    "        if alignedSeq[alignCounter]==\"-\":\n",
    "            newUlist.append(0)\n",
    "            alignCounter=alignCounter+1\n",
    "        else:\n",
    "            newUlist.append(ulist[oldUcounter])\n",
    "            oldUcounter=oldUcounter+1\n",
    "            alignCounter=alignCounter+1\n",
    "            if (oldUcounter+1)==len(ulist):\n",
    "                newUlist.append(ulist[oldUcounter])\n",
    "    return newUlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 1, 0, 1, 0, 0, 0, 5, 0, 3, 0, 3, 2, 1, 0, 0, 0, 0, 0, 1, 1, 2, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 1, 0, 0, 2, 2, 1, 0, 1, 1, 4, 4, 3, 2, 2, 1, 1, 3, 0, 5, 1, 1, 1, 3, 0, 2, 2, 2, 4, 2, 3, 1, 2, 0, 2, 1, 2, 1, 0, 0, 2, 3, 1, 2, 1, 1, 3, 0, 0, 0, 1, 0, 0, 0, 0]\n",
      "[0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 5, 0, 3, 0, 3, 2, 1, 0, 0, 0, 0, 0, 1, 1, 2, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 1, 0, 0, 2, 2, 1, 0, 1, 1, 4, 4, 3, 2, 2, 1, 1, 3, 0, 5, 1, 1, 1, 3, 0, 2, 2, 2, 4, 2, 3, 1, 2, 0, 2, 1, 2, 1, 0, 0, 2, 3, 1, 2, 1, 1, 3, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "ND3updatedUs=list()\n",
    "print ND3alignedShort[3].uList\n",
    "for i in range(len(ND3alignedShort)):\n",
    "    ND3updatedUs.append(ND3alignedShort[i])\n",
    "    ND3updatedUs[i].uList=updateUs(ND3alignedShort[i].uList,ND3alignedShort[i].noUs)\n",
    "print ND3updatedUs[3].uList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# masterUlistmaker function\n",
    "### This function creates a master U list that contains the highest number of Us observed at every edited position.  It makes a dummy list of all of the different numbers of Us at a given position and then records the maximum from the dummy list, and then moves on the the next position.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def masterUlistmaker(seqList):\n",
    "    #make a master list of most abundant Us\n",
    "    masterUList=list()\n",
    "\n",
    "    # make a list containing all of the numbers of Us at position i\n",
    "    for i in range(len(seqList[0].uList)):\n",
    "        dummyUlist=list()\n",
    "        for j in range(len(seqList)):\n",
    "            dummyUlist.append(seqList[j].uList[i])\n",
    "        #record the highest number in the master list\n",
    "        masterUList.append(max(dummyUlist))\n",
    "    return masterUList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 5, 0, 3, 0, 3, 2, 1, 0, 2, 5, 1, 3, 5, 1, 2, 2, 1, 0, 3, 4, 4, 3, 1, 0, 0, 9, 1, 6, 1, 6, 2, 0, 0, 7, 0, 2, 4, 0, 2, 2, 3, 1, 0, 4, 0, 1, 1, 0, 3, 1, 1, 0, 1, 1, 0, 0, 2, 0, 1, 0, 3, 1, 1, 4, 1, 0, 0, 2, 0, 0, 3, 1, 4, 2, 0, 2, 3, 6, 1, 2, 2, 2, 2, 2, 0, 2, 2, 4, 4, 5, 2, 3, 5, 5, 5, 1, 5, 2, 2, 5, 5, 2, 3, 5, 7, 7, 5, 4, 5, 6, 3, 3, 3, 2, 3, 3, 4, 2, 3, 1, 2, 1, 1, 5, 0, 0, 0, 1, 0, 0, 0, 1, 1, 3, 3, 5, 0, 1, 1, 0, 4, 4, 3, 1, 0, 1, 6, 2, 1, 2, 1, 1, 0, 1, 2, 1, 2, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "ND3masterU=masterUlistmaker(ND3updatedUs)\n",
    "print ND3masterU"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# putTheUsBack function\n",
    "### This function reinserts the correct number of Us back into the noUs sequence, and it inserts dashes if the number of Us at a given position in a sequence is less that the highest number observed at that position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def putTheUsBack(align1,newulist1,masterUlist):\n",
    "    \n",
    "    newalign1=\"\"\n",
    "\n",
    "    for i in range(len(align1)):\n",
    "        #insert the number of Us specified if it is the same or greater than those specified in the masterUlist\n",
    "        if newulist1[i]>=masterUlist[i]:\n",
    "            newalign1=newalign1+\"U\"*newulist1[i]\n",
    "        #insert dashes to fill space\n",
    "        else:\n",
    "            newalign1=newalign1+\"-\"*(masterUlist[i]-newulist1[i])+\"U\"*newulist1[i]\n",
    "        newalign1=newalign1+align1[i]\n",
    "    #This checks the last position as the uList is always one item longer than the alignment string\n",
    "    if newulist1[len(align1)]>=masterUlist[len(align1)]:\n",
    "        newalign1=newalign1+\"U\"*newulist1[len(align1)]\n",
    "    else:\n",
    "        newalign1=newalign1+\"-\"*(masterUlist[len(align1)]-newulist1[len(align1)])+\"U\"*newulist1[len(align1)]\n",
    "\n",
    "    return newalign1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCAAAAAAUCCUCGCCUUUUUACUUUAGUUUGUUAUCAUUAUUUUUAUAUUUGUUUUUG-A-UAUUGUGGUUUA--UUAUUUUAUUUAUAGGUUUUUUUUUAUGUUUUUUAUGUUUUUUAUUGCAUUUUUUUGAUUGUUUUCGUUGUUGUUUGUGGUUUUCGUGUGGUUUGUAUGAUAUGAAUUCA-CGUUUG-GUGUUUUAUACAUUGGAUUUAUGUUUUGUUAGUUGUUUGUUUUUUGUAUUGUUA--A--AUUCC--AUUA-UUUG---UG-UUUUGUUGUUUGUUUUUG----UG-----AUA-----G-UGUUG-UUUUAUUUUUGUUA--UG-----G-UUUUUUG--UUUUUG----UG----GUUUUUGUUUUUUG-UUG--UA--UG-UA--UA---G----G--AUUUGUG-UG-GUAUUUUUGGGAUCACGUAUAUUUG--UG----UGGUGUAAUUUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUAUA-CAUAUUAUAUUAAUAAAUAUAUAAAA\n",
      "--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------UUUUUG--UUUUUG-------G--UUUG--UUG-UUUUG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--AUUUGUGUUGUGUA--UUUGGGAUCACGUAUA-UUGUUUGUUUUUGGUG-AA-UUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUAUAUA-------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A--A--A--CCUUA--A----G--UUG---UUGUUA-UUG----UG----UG-----AUA----UG--G-UGUUUUUA----UG-UAUUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG------G-UUGUUU------------------------------------------------------------------------------------------------------------------------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---A-G----G--AG-UG---G----UUGUA--G--AUUAUUA-UCC-UA-UAUUUUGUUUUG--UUUGUUA-UUG----UG----UG--UUUA-AUUUUUG-UG-UG----UA--UUUG--A-UUG---UUG-----UUG---UUUUG---UUG-UUUG----UG----UUG---G-UUA--UGUUA--UA---G----GUUAUUUGUGUUGUGUA--UUUGGGAUCACG---------------------------------------------------------------------------------------\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------G---UUUUG-----G----G-----G------G---G---A---G--A---A---GUUUUG--AUUUGUGUUGUGUA--UUUGGGAUCACGUAUA-UUG--UG----UGGUG-AA-UUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUAUAUA-------------------------\n",
      "------AAUCCUCGCC-UUUUACUUUAGUUUGUUAUCA--A-----G-A---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------G-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A--A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG-----A-A-----G--G-UGUUUUUA----UG-UAUUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--A------------------------------------------------------------------------------------------------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-G---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A--A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG-----A-A----UG--G--G-----A--UUUG--A-UUG---UUG-------GUUUUUUUGUUUUUG--UUG-UUUUG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--A------------------------------------------------------------------------------------------------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---AUG----G--GG-UG-UUG------G-A--G--A--A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG-----A-A----UG--G-UGUUUUUA----UG-UAUUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--A------------------------------------------------------------------------------------------------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-G---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCAUGG--UGUG------A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A--A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG-----A-A-----G--G-UGUUUUUA-UUUUG--A--UG-----G--UUUUUG-------G-UUUUG----G----UG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--A------------------------------------------------------------------------------------------------------------------\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A---G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------A-G------A-G------G--GCA-------GA--G----CG--G--G---G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A--A--A--CC--A--A----G----GUUUUUGUUA--UG-----GUUUUUGUUUUUAUA---UUGUUG-UG-----A--UUUGUUA-UUG-----GUUUUUUUG--UUUUUG-----GUUUUG--UUUG---UUUGUUUG---AUUUG-UAUUUAUUUG--UUG--A------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "ND3shortFinal=list()\n",
    "for i in range(len(ND3updatedUs)):\n",
    "    ND3shortFinal.append(ND3updatedUs[i])\n",
    "    ND3shortFinal[i].alignedSeq=putTheUsBack(ND3shortFinal[i].noUs,ND3shortFinal[i].uList,ND3masterU)\n",
    "    print ND3shortFinal[i].alignedSeq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# lineFit function\n",
    "### This function rewrites the aligned sequences in a more visually appealing fashion within a single string.  It takes the final aligned sequences and records a certain number of characters per line, with their titles shown at the end of each line.  It also prints a title supplied by the user."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def lineFit(seqList,alignmentName,lineMax):\n",
    "    alignLength=len(seqList[0].alignedSeq)\n",
    "    alignment=\">\"+alignmentName+\"\\n\"\n",
    "    start=0 #counter that shows where to start copying for each new line\n",
    "    #copy down a set number of characters from each alignment\n",
    "    #repeat until what is left is shorter than the specified max line length\n",
    "    while alignLength>lineMax:\n",
    "        for j in range(len(seqList)):\n",
    "            for i in range(start,(start+lineMax)):\n",
    "                alignment=alignment+seqList[j].alignedSeq[i]\n",
    "            alignment=alignment+\"-\"+seqList[j].title+\"\\n\"\n",
    "        alignment=alignment+\"\\n\"\n",
    "        start=start+lineMax\n",
    "        alignLength=alignLength-lineMax\n",
    "    #copy down the remaining characters from each alignment\n",
    "    for j in range(len(seqList)):\n",
    "        for i in range(start,(start+alignLength)):\n",
    "            alignment=alignment+seqList[j].alignedSeq[i]\n",
    "        alignment=alignment+\"-\"+seqList[j].title+\"\\n\"\n",
    "\n",
    "    return alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ND3 short alignment\n",
      "UCAAAAAAUCCUCGCCUUUUUACUUUAGUUUGUUAUCAUUAUUUUUAUAU-ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A--ND3AltA02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A--ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "------AAUCCUCGCC-UUUUACUUUAGUUUGUUAUCA--A-----G-A--ND3Alt2C02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-G--ND3AltD02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A--ND3AltE02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-G--ND3Alt3B01\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUAUCA--A-----A-A--ND3Alt3B02\n",
      "\n",
      "UUGUUUUUG-A-UAUUGUGGUUUA--UUAUUUUAUUUAUAGGUUUUUUUU-ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3AltA02\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3Alt2C02\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3AltD02\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3AltE02\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3Alt3B01\n",
      "--G----UGUAUUA--G-GG---AUUUUA----A---A-AGG---------ND3Alt3B02\n",
      "\n",
      "UAUGUUUUUUAUGUUUUUUAUUGCAUUUUUUUGAUUGUUUUCGUUGUUGU-ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3AltA02\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "-G-G------A-G------G--GCA-------GA--G----CG--G--G--ND3Alt2C02\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3AltD02\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3AltE02\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3Alt3B01\n",
      "-A-G------A-G------G--GCA-------GA--G----CG--G--G--ND3Alt3B02\n",
      "\n",
      "UUGUGGUUUUCGUGUGGUUUGUAUGAUAUGAAUUCA-CGUUUG-GUGUUU-ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3AltA02\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3Alt2C02\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3AltD02\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3AltE02\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCAUGG--UGUG------ND3Alt3B01\n",
      "--G-GA----CG-G-GG-UUG-A-GA-A-GAA-UCA-CG---GUGUG----ND3Alt3B02\n",
      "\n",
      "UAUACAUUGGAUUUAUGUUUUGUUAGUUGUUUGUUUUUUGUAUUGUUA---ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "-A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A---ND3AltA02\n",
      "-A-ACA--GGA---A-G----G--AG-UG---G----UUGUA--G--AUU-ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "-A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A---ND3Alt2C02\n",
      "-A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A---ND3AltD02\n",
      "-A-ACA--GGA---AUG----G--GG-UG-UUG------G-A--G--A---ND3AltE02\n",
      "-A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A---ND3Alt3B01\n",
      "-A-ACA--GGA---AUG----G--AG-UG-UUG------G-A--G--A---ND3Alt3B02\n",
      "\n",
      "A--AUUCC--AUUA-UUUG---UG-UUUUGUUGUUUGUUUUUG----UG--ND3Conv\n",
      "---------------------------------------------------ND3AltF01\n",
      "A--A--CCUUA--A----G--UUG---UUGUUA-UUG----UG----UG--ND3AltA02\n",
      "AUUA-UCC-UA-UAUUUUGUUUUG--UUUGUUA-UUG----UG----UG--ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG--ND3Alt2C02\n",
      "A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG--ND3AltD02\n",
      "A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG--ND3AltE02\n",
      "A--A--CC--A-UA----G----G--UUUG--A---G-----G----UG--ND3Alt3B01\n",
      "A--A--CC--A--A----G----GUUUUUGUUA--UG-----GUUUUUGU-ND3Alt3B02\n",
      "\n",
      "----AUA-----G-UGUUG-UUUUAUUUUUGUUA--UG-----G-UUUUU-ND3Conv\n",
      "--------------------------------------UUUUUG--UUUU-ND3AltF01\n",
      "----AUA----UG--G-UGUUUUUA----UG-UAUUUG--UUUG-------ND3AltA02\n",
      "-UUUA-AUUUUUG-UG-UG----UA--UUUG--A-UUG---UUG-----U-ND3Alt4E01\n",
      "---------------------------------------------------ND3Alt2A03\n",
      "----A-A-----G--G-UGUUUUUA----UG-UAUUUG--UUUG-------ND3Alt2C02\n",
      "----A-A----UG--G--G-----A--UUUG--A-UUG---UUG-------ND3AltD02\n",
      "----A-A----UG--G-UGUUUUUA----UG-UAUUUG--UUUG-------ND3AltE02\n",
      "----A-A-----G--G-UGUUUUUA-UUUUG--A--UG-----G--UUUU-ND3Alt3B01\n",
      "UUUUAUA---UUGUUG-UG-----A--UUUGUUA-UUG-----GUUUUUU-ND3Alt3B02\n",
      "\n",
      "UG--UUUUUG----UG----GUUUUUGUUUUUUG-UUG--UA--UG-UA--ND3Conv\n",
      "UG-------G--UUUG--UUG-UUUUG---UUUGUUUG---AUUUG-UAU-ND3AltF01\n",
      "UG---UUUUG-UUUUG---UG---UUG------G-UUGUUU----------ND3AltA02\n",
      "UG---UUUUG---UUG-UUUG----UG----UUG---G-UUA--UGUUA--ND3Alt4E01\n",
      "-G---UUUUG-----G----G-----G------G---G---A---G--A--ND3Alt2A03\n",
      "UG---UUUUG-UUUUG---UG---UUG---UUUGUUUG---AUUUG-UAU-ND3Alt2C02\n",
      "-GUUUUUUUGUUUUUG--UUG-UUUUG---UUUGUUUG---AUUUG-UAU-ND3AltD02\n",
      "UG---UUUUG-UUUUG---UG---UUG---UUUGUUUG---AUUUG-UAU-ND3AltE02\n",
      "UG-------G-UUUUG----G----UG---UUUGUUUG---AUUUG-UAU-ND3Alt3B01\n",
      "UG--UUUUUG-----GUUUUG--UUUG---UUUGUUUG---AUUUG-UAU-ND3Alt3B02\n",
      "\n",
      "-UA---G----G--AUUUGUG-UG-GUAUUUUUGGGAUCACGUAUAUUUG-ND3Conv\n",
      "UUAUUUG--UUG--AUUUGUGUUGUGUA--UUUGGGAUCACGUAUA-UUG-ND3AltF01\n",
      "---------------------------------------------------ND3AltA02\n",
      "-UA---G----GUUAUUUGUGUUGUGUA--UUUGGGAUCACG---------ND3Alt4E01\n",
      "--A---GUUUUG--AUUUGUGUUGUGUA--UUUGGGAUCACGUAUA-UUG-ND3Alt2A03\n",
      "UUAUUUG--UUG--A------------------------------------ND3Alt2C02\n",
      "UUAUUUG--UUG--A------------------------------------ND3AltD02\n",
      "UUAUUUG--UUG--A------------------------------------ND3AltE02\n",
      "UUAUUUG--UUG--A------------------------------------ND3Alt3B01\n",
      "UUAUUUG--UUG--A------------------------------------ND3Alt3B02\n",
      "\n",
      "--UG----UGGUGUAAUUUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUA-ND3Conv\n",
      "UUUGUUUUUGGUG-AA-UUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUA-ND3AltF01\n",
      "---------------------------------------------------ND3AltA02\n",
      "---------------------------------------------------ND3Alt4E01\n",
      "--UG----UGGUG-AA-UUUAUUUUGUUUAUGAUGUUUUUUGUUGUAUUA-ND3Alt2A03\n",
      "---------------------------------------------------ND3Alt2C02\n",
      "---------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------ND3AltE02\n",
      "---------------------------------------------------ND3Alt3B01\n",
      "---------------------------------------------------ND3Alt3B02\n",
      "\n",
      "UA-CAUAUUAUAUUAAUAAAUAUAUAAAA-ND3Conv\n",
      "UAUA--------------------------ND3AltF01\n",
      "------------------------------ND3AltA02\n",
      "------------------------------ND3Alt4E01\n",
      "UAUA--------------------------ND3Alt2A03\n",
      "------------------------------ND3Alt2C02\n",
      "------------------------------ND3AltD02\n",
      "------------------------------ND3AltE02\n",
      "------------------------------ND3Alt3B01\n",
      "------------------------------ND3Alt3B02\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ND3shortAlignment=lineFit(ND3shortFinal,\"ND3 short alignment\",50)\n",
    "print ND3shortAlignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# multipleEdSeqAlign function\n",
    "### This function ties all of the previous functions together.  It extracts the sequences from a file using extractSequences, then it aligns the sequences using alignAll. Then it updates their u lists using updateUs and it generates a master u list using masterUlistmaker.  Finally, it reinserts the Us into the aligned sequences using putTheUsBack and creates an alignment using lineFit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def multipleEdSeqAlign(filename,alignmentName,lineMax):\n",
    "    allSeq=extractSequences(filename)\n",
    "    updatedSeq=alignAll(allSeq)\n",
    "    for i in range(len(updatedSeq)):\n",
    "        updatedSeq[i].uList=updateUs(updatedSeq[i].uList,updatedSeq[i].noUs)      \n",
    "    masterUlist=masterUlistmaker(updatedSeq)\n",
    "    for i in range(len(updatedSeq)):\n",
    "        updatedSeq[i].alignedSeq=putTheUsBack(updatedSeq[i].noUs,updatedSeq[i].uList,masterUlist)\n",
    "    alignment=lineFit(updatedSeq,alignmentName,lineMax)\n",
    "    return alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ND3alternativealignment\n",
      "UCAAAAAAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCAUUA-UUUUUAUAUUUGUUUUUG-A----UAUUG--UG--GU-ND3ConvED\n",
      "UCAAAAAAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Uned\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3AltA02\n",
      "---------------------------------------------------------------------------------ND3Alt2A03\n",
      "------AAUCCUCGCC-UUUUACUUUAGUUUGUUA-UCA--A------G-A---G----UGUA---UUA--G---G--G--ND3AltC02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-G---G----UGUA---UUA--G---G--G--ND3AltD02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3AltE02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-G---G----UGUA---UUA--G---G--G--ND3Alt3B01\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt3B02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt3D01\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt3D09\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUG--AUUCA--AUUUUUUAUA---G---UUG-AUUUUUA-UGUUUGUUG--ND3Alt3E05\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt3G03\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt3G07\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt4B02\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt4E01\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt4F04\n",
      "------AAUCCUCGCCUUUUUACUUUAGUUUGUUA-UCA--A------A-A---G----UGUA---UUA--G---G--G--ND3Alt4G02\n",
      "---------------------------------------------------------------------------------ND3AltC01\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltD03\n",
      "---------------------------------------------------------------------------------ND3AltF01\n",
      "---------------------------------------------------------------------------------ND3AltF03\n",
      "---------------------------------------------------------------------------------ND3AltG01\n",
      "\n",
      "UUA--UUAUUUUAUUUA-UA----GGUUUUUUUUUAUGUUUUUUA---UGUUUUUUAUUG-C-AUUUUUUUG---A-UUG-ND3ConvED\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------A--G-C-A-------G---A---G-ND3Uned\n",
      "--AUUUUA----A---A--A--G-G----------A-G------A----G------G--G-C-A-------G---A---G-ND3AltA02\n",
      "---------------------------------------------------------------------------------ND3Alt2A03\n",
      "--AUUUUA----A---A--A----GG---------G-G------A----G------G--G-C-A-------G---A---G-ND3AltC02\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3AltD02\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3AltE02\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3B01\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3B02\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3D01\n",
      "--A---UC--UUA---A--A--A-GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3D09\n",
      "UUA---UA---UAUUUAUUAUUGUG---------UAUG------AUUUUG------G--GUCUA-------GUUUAUUUG-ND3Alt3E05\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3G03\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt3G07\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt4B02\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt4E01\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt4F04\n",
      "--AUUUUA----A---A--A----GG---------A-G------A----G------G--G-C-A-------G---A---G-ND3Alt4G02\n",
      "---------------------------------------------------------------------------------ND3AltC01\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltD03\n",
      "---------------------------------------------------------------------------------ND3AltF01\n",
      "---------------------------------------------------------------------------------ND3AltF03\n",
      "---------------------------------------------------------------------------------ND3AltG01\n",
      "\n",
      "UUUUCGUUG-UUGUUUGUG----G-UUUUCGUGUGGUUUG--UA-UG----AUA-UGA--AUUC--A-CGUUUG-GUGUU-ND3ConvED\n",
      "----CG--G---G---G-G----G-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Uned\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3AltA02\n",
      "---------------------------------------------------------------------------------ND3Alt2A03\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3AltC02\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3AltD02\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3AltE02\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--AUGG--UGUG-----ND3Alt3B01\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt3B02\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt3D01\n",
      "----CG--G---G---G-G----A-----CG-G-GG---G--UA-UGUUUUAUA-UGAUUA-UCUUA-CG---GUGUG---ND3Alt3D09\n",
      "-UUUCGUUGUUUG-UUGUGUUUUCA----CG-GUGG---GUUUAUUG-UUUAUAUUGAUUA--C--A-CG-UUG-G-G---ND3Alt3E05\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt3G03\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt3G07\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt4B02\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt4E01\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt4F04\n",
      "----CG--G---G---G-G----A-----CG-G-GG-UUG---A--G----A-A--GA--A-UC--A-CG---GUGUG---ND3Alt4G02\n",
      "---------------------------------------------------------------------------------ND3AltC01\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltD03\n",
      "---------------------------------------------------------------------------------ND3AltF01\n",
      "---------------------------------------------------------------------------------ND3AltF03\n",
      "---------------------------------------------------------------------------------ND3AltG01\n",
      "\n",
      "UUAUA--CA-UUG---G-----AUUUA-UGUUUUG--UUA---G-UUGUUUGUUUUUUG--------UA-UUG--UUA---ND3ConvED\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Uned\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G----A--------G----A---ND3AltA02\n",
      "---------------------------------------------------------------------------------ND3Alt2A03\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3AltC02\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3AltD02\n",
      "--A-A--CA---G---G-----A---A-UG----G----G---G--UG-UUG------G---------A---G----A---ND3AltE02\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt3B01\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt3B02\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG----UUG---------A---G--UUA-U-ND3Alt3D01\n",
      "--A-AUUCAUUUG--UGUUUUUAUUUA--GUUUUGUUUUA-UUGUUUG---C------GUUUUGUUUUAUUUGUUUUA-U-ND3Alt3D09\n",
      "-UA-A-UCA-UUGUUUG---UUA-UUAUUG----G----AUUUG-UUG--UG-----UG----A------UUG----A---ND3Alt3E05\n",
      "--A-AUUCA--UGUUUG----UA-UUA-UG----G--UUA-UUG-UUG---G------G-------UUA-UUG---UA-U-ND3Alt3G03\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt3G07\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt4B02\n",
      "--A-A--CA---G---G-----A---A--G----G----A---G--UG---G----UUG--------UA---G----AUU-ND3Alt4E01\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt4F04\n",
      "--A-A--CA---G---G-----A---A-UG----G----A---G--UG-UUG------G---------A---G----A---ND3Alt4G02\n",
      "---------------------------------------------------------------------------------ND3AltC01\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltD03\n",
      "---------------------------------------------------------------------------------ND3AltF01\n",
      "---------------------------------------------------------------------------------ND3AltF03\n",
      "---------------------------------------------------------------------------------ND3AltG01\n",
      "\n",
      "A-----AUUCC--AUUA----UUUG----UG-UUUUG--UUGUUUGUUUUUG----UG-----A--UA-------G--UG-ND3ConvED\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----G---G-----G----UG-----A---A------UG---G-ND3Uned\n",
      "A-----A--CCUUA--A-------G---UUG---UUG--UUA-UUG----UG----UG-----A--UA------UG---G-ND3AltA02\n",
      "---------------------------------------------------------------------------------ND3Alt2A03\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A-------G---G-ND3AltC02\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A------UG---G-ND3AltD02\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A------UG---G-ND3AltE02\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A-------G---G-ND3Alt3B01\n",
      "A-----A--CC--A--A-------G-----GUUUUUG--UUA--UG-----GUUUUUGUUUUUA--UA-----UUG-UUG-ND3Alt3B02\n",
      "A----UA--CC--A-UA------UG--UUUG-----G--UUA--UG-----GUUUUUGUUUUUA--UA-----UUG-UUG-ND3Alt3D01\n",
      "A---UUA--CC--AUUA-------G-----G--UUUG--UUAUUUG-----GUUUUUGUUUUUA--UA-------G---G-ND3Alt3D09\n",
      "A---UUG--CC--A--AUUUUUUUGUUUUUG-----GUUUUAUUUG--UUUG--UUUG-----AUUUGUA--UUUAUUUG-ND3Alt3E05\n",
      "AUUUUUA-UCC--A--A-------G---UUG---UUG--UUA-UUG----UG-----G-----A--UA-----UUG-UUG-ND3Alt3G03\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A-------G---G-ND3Alt3G07\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A------UG---G-ND3Alt4B02\n",
      "A---UUA-UCC-UA-UA---UUUUG-UUUUG--UUUG--UUA-UUG----UG----UG--UUUA---A--UUUUUG--UG-ND3Alt4E01\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG-----A---A-------G---G-ND3Alt4F04\n",
      "A-----A--CC--A-UA-------G-----G--UUUG----A---G-----G----UG----UA---A------UG---G-ND3Alt4G02\n",
      "---------------------------------------------------------------------------------ND3AltC01\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltD03\n",
      "---------------------------------------------------------------------------------ND3AltF01\n",
      "---------------------------------------------------------------------------------ND3AltF03\n",
      "---------------------------------------------------------------------------------ND3AltG01\n",
      "\n",
      "---UUG-UUUUAUUUUUG----UUA----UG-----G-UUUUUUG--UUUUUG----UG----GUUUUUGUUUUUUG-UU-ND3ConvED\n",
      "-----G-----A-----G------A----UG-----G-------G---UUUUG-----G----G-----G------G----ND3Uned\n",
      "----UGUUUUUA----UG-----UA--UUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG------G-UU-ND3AltA02\n",
      "------------------------------G-UUUUG-------G-------G-----G----G-----G-----------ND3Alt2A03\n",
      "----UGUUUUUA----UG-----UA--UUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG---UUUGUUU-ND3AltC02\n",
      "-----G-----A--UUUG------A---UUG---UUG-------GUUUUUUUGUUUUUG--UUG-UUUUG---UUUGUUU-ND3AltD02\n",
      "----UGUUUUUA----UG-----UA--UUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG---UUUGUUU-ND3AltE02\n",
      "----UGUUUUUA-UUUUG------A----UG-----G--UUUUUG-------G-UUUUG----G----UG---UUUGUUU-ND3Alt3B01\n",
      "----UG-----A--UUUG----UUA---UUG-----GUUUUUUUG--UUUUUG-----GUUUUG--UUUG---UUUGUUU-ND3Alt3B02\n",
      "----UG-----A--UUUG----UUA---UUG-----GUUUUUUUG--UUUUUG-----GUUUUG--UUUG---UUUGUUU-ND3Alt3D01\n",
      "----UGUUUUUA----UG-----UA--UUUG-----A-----UUG------UG-UUUUGUUUUG----UG-----UG----ND3Alt3D09\n",
      "---UUG-----A---------------------------------------------------------------------ND3Alt3E05\n",
      "UUUUUG----UA----UG---UUUA--UUUG----UG---UUUUG---UUUUG----UG--UUG--UUUG---UUUG----ND3Alt3G03\n",
      "----UGUUUUUA----UG-----UA--UUUG--UUUG------UG---UUUUG-UUUUG---UG---UUG---UUUGUUU-ND3Alt3G07\n",
      "-----G-----A-----G------A---UUG----UG------UG----UUUG---UUG-UUUG----UG----UUG----ND3Alt4B02\n",
      "----UG----UA--UUUG------A---UUG---UUG-----UUG---UUUUG---UUG-UUUG----UG----UUG----ND3Alt4E01\n",
      "---UUG--UUUA----UG-UUUUUA----UG-UUUUG---UUUUG----UUUG---UUG-UUUG----UG----UUG----ND3Alt4F04\n",
      "-----G-----A-----GUUUUUUA--UUUG----UG----UUUG----UUUG---UUG-UUUG----UG----UUG----ND3Alt4G02\n",
      "-----------------------------UG--UUUG------UG-----UUG-----G----------------------ND3AltC01\n",
      "------------------------------G----UG----UUUG-----UUG--UUUG---UG---UUG------G----ND3AltD02\n",
      "----------------------------UUG---UUG----UUUG------UG---UUG----G-----------------ND3AltD03\n",
      "-------------------------UUUUUGUUUUUG-------G----UUUG---UUGUUUUG--UUUG---UUUG----ND3AltF01\n",
      "------------------------------G-UUUUG-------G-------G-----G----G-----G-----------ND3AltF03\n",
      "--------------------------UUUUG-UUUUG-----UUG----UUUG----UGUUUUG-----G-----------ND3AltG01\n",
      "\n",
      "G------UA------UG-UA---UA---G----G---AUUUG--------UG-UG--G-UAUUUUUG---G-G--A--UC-ND3ConvED\n",
      "G-------A-------G--A----A---GUUUUG---A--UG---------G--G--G--A-----G---G-G--A--UC-ND3Uned\n",
      "GUUU-----------------------------------------------------------------------------ND3AltA02\n",
      "----A-------G------A----A---GUUUUG---AUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3Alt2A03\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3AltC02\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3AltD02\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3AltE02\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3B01\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3B02\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3D01\n",
      "GUUUGUUUG---AUUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3D09\n",
      "---------------------------------------------------------------------------------ND3Alt3E05\n",
      "A------------UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3G03\n",
      "G-------A----UUUG-UA-UUUAUUUG--UUG---A-------------------------------------------ND3Alt3G07\n",
      "G-----UUA------UGUUA---UA---G----G---AUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3Alt4B02\n",
      "G-----UUA------UGUUA---UA---G----G-UUAUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3Alt4E01\n",
      "G-----UUA------UGUUA---UA---G----G-UUAUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3Alt4F04\n",
      "G-----UUA------UGUUA---UA---G----G-UUAUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3Alt4G02\n",
      "--UUA------UG----UUAUUUUA---G---UG---AUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3AltC01\n",
      "--UUA------UG----UUA---UA---G----G-UUAUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3AltD02\n",
      "--UUA------UG----UUA----A---G----G---AUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3AltD03\n",
      "----A----UUUG-----UA-UUUAUUUG--UUG---AUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3AltF01\n",
      "----A-------G------A--UUA---G--UUGUUUA---GUUUUUUUUUG--GUUGUUA----UGUUUGUGUUAUUUC-ND3AltF03\n",
      "---UA-----UUG-----UA---UA-UUG---UG--UAUUUG--------UGUUG-UG-UA--UUUG---G-G--A--UC-ND3AltG01\n",
      "\n",
      "AC-GUAUA--UUUG--UG----UG-----GUGUA--AUUUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUA-CAUA-ND3ConvED\n",
      "ACUG-A-A-----G---G-----G-----G-GUA--A----A----G---A-GAUUUG---UUUGUUGUAUUAUA-CAUA-ND3Uned\n",
      "---------------------------------------------------------------------------------ND3AltA02\n",
      "AC-GUAUA---UUG--UG----UG-----GUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3Alt2A03\n",
      "---------------------------------------------------------------------------------ND3AltC02\n",
      "---------------------------------------------------------------------------------ND3AltD02\n",
      "---------------------------------------------------------------------------------ND3AltE02\n",
      "---------------------------------------------------------------------------------ND3Alt3B01\n",
      "---------------------------------------------------------------------------------ND3Alt3B02\n",
      "---------------------------------------------------------------------------------ND3Alt3D01\n",
      "---------------------------------------------------------------------------------ND3Alt3D09\n",
      "---------------------------------------------------------------------------------ND3Alt3E05\n",
      "---------------------------------------------------------------------------------ND3Alt3G03\n",
      "---------------------------------------------------------------------------------ND3Alt3G07\n",
      "AC-G-----------------------------------------------------------------------------ND3Alt4B02\n",
      "AC-G-----------------------------------------------------------------------------ND3Alt4E01\n",
      "AC-G-----------------------------------------------------------------------------ND3Alt4F04\n",
      "AC-G-----------------------------------------------------------------------------ND3Alt4G02\n",
      "AC-G-AUA----UG--UG----UG-----GUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltC01\n",
      "AC-GUAUA---UUG--UG----UG-----G-G-AUUA-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltD02\n",
      "AC-GUAUA--UUUG--UG---UUG-----GUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltD03\n",
      "AC-GUAUA---UUGUUUGUUUUUG-----GUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltF01\n",
      "AC-GUA-AUUUUUG-UUGUUUUUGUUUUUGUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltF03\n",
      "AC-GUAUA---UUGUUUGUUUUUG-----GUG-A--A-UUUAUUUUGUUUAUGA--UGUUUUUUGUUGUAUUAUAUA----ND3AltG01\n",
      "\n",
      "UUAUAUUAAUAAAUAUAUAAAA--------------ND3ConvED\n",
      "UUAUAUUAAUAAAUAUAUAAAAUAAUAAAAUACUA-ND3Uned\n",
      "------------------------------------ND3AltA02\n",
      "------------------------------------ND3Alt2A03\n",
      "------------------------------------ND3AltC02\n",
      "------------------------------------ND3AltD02\n",
      "------------------------------------ND3AltE02\n",
      "------------------------------------ND3Alt3B01\n",
      "------------------------------------ND3Alt3B02\n",
      "------------------------------------ND3Alt3D01\n",
      "------------------------------------ND3Alt3D09\n",
      "------------------------------------ND3Alt3E05\n",
      "------------------------------------ND3Alt3G03\n",
      "------------------------------------ND3Alt3G07\n",
      "------------------------------------ND3Alt4B02\n",
      "------------------------------------ND3Alt4E01\n",
      "------------------------------------ND3Alt4F04\n",
      "------------------------------------ND3Alt4G02\n",
      "------------------------------------ND3AltC01\n",
      "------------------------------------ND3AltD02\n",
      "------------------------------------ND3AltD03\n",
      "------------------------------------ND3AltF01\n",
      "------------------------------------ND3AltF03\n",
      "------------------------------------ND3AltG01\n",
      "\n",
      "47.3899998665\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start=time.time()\n",
    "print multipleEdSeqAlign(\"ND3align3.fa\",\"ND3alternativealignment\",80)\n",
    "end=time.time()\n",
    "print (end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">CR3alternativealignment\n",
      "AGAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAAUGUUUGA----UUG-UUUG----GUUUUG----UUGUUUUUUUA--CR3Conv001\n",
      "---------------------------------------------------------------------------------CR3AltB01\n",
      "---------------------------------------------------------------------------------CR3AltG01\n",
      "---------------------------------------------------------------------------------CR3AltH01\n",
      "---------------------------------------------------------------------------------CR3AltA02\n",
      "---------------------------------------------------------------------------------CR3AltB02\n",
      "---------------------------------------------------------------------------------CR3AltD01\n",
      "---------------------------------------------------------------------------------CR3AltC01\n",
      "---------------------------------------------------------------------------------CR3AltE01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA----UUG--UUGUUUUGUUUUGUUUUUUA--------U-CR3Alt1D01\n",
      "AGAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAAUG---GA-----UG----G----G----G------G----UUUA--CR3Alt1G01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA-----UGUUUUG----G---UG------G----UUUA--CR3Alt1H01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA----UUG----G----G----G------G----UUUA--CR3Alt2A01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GAUUUUUUG----G----G----G------G------UA--CR3Alt2B01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GAUUUUUUG----G----G---UG------G----------CR3Alt2C01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA----UUG----G----G----G------G----UUUA--CR3Alt2D01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA---UUUGUUUUG----G---UG------G----UUUA--CR3Alt2E01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GAUUUUUUG----G----G----G------G----UUUA--CR3Alt2G01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA----UUG----G----G----G------G-----UUA--CR3Alt2H01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GA------G---UG----G----G------G----UUUA--CR3Alt2A02\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2H02\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-G---GAUUUUUUG----G----G----G------G----UUUA--CR3Alt2B03\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2C03\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2E03\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2F03\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2A04\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt2C04\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt3A01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt3B01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt3G01\n",
      "-GAAAUAUAAAUAUGUGUAUGAUAUAUAAAAACAA-A---GA----UUG----G----G----G------G----UUUA--CR3Alt3B02\n",
      "\n",
      "UUGUUUGUUUGUA---C-----AUUUUUUUUGUUUUUUAUUUG-UUUGUG-AUUUGUUUUUAUGUUUGUUAUUUAGUUUU-CR3Conv001\n",
      "---------------------------------------------------------------------------------CR3AltB01\n",
      "---------------------------------------------------------------------------------CR3AltG01\n",
      "---------------------------------------------------------------------------------CR3AltH01\n",
      "---------------------------------------------------------------------------------CR3AltA02\n",
      "---------------------------------------------------------------------------------CR3AltB02\n",
      "---------------------------------------------------------------------------------CR3AltD01\n",
      "---------------------------------------------------------------------------------CR3AltC01\n",
      "---------------------------------------------------------------------------------CR3AltE01\n",
      "UUGUUUG--UG-AUUUGUUUUUA-------UG---UUUG------------------------------------------CR3Alt1D01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt1G01\n",
      "UUG--UG---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt1H01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2A01\n",
      "-UG--UG---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2B01\n",
      "---------------------------------------------------------------------------------CR3Alt2C01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2D01\n",
      "UUG--UG--UGUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2E01\n",
      "UUG---G---GUA--UC-----A--------A-----UAUUUGUUUUG-AGAUUUGUUUUUAUG-----------------CR3Alt2G01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2H01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2A02\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2H02\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2B03\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2C03\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2E03\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2F03\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2A04\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt2C04\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt3A01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt3B01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt3G01\n",
      "--G---G---GUA--UC-----A--------G-----UAUUUGUUUUGUG-AUUUGUUUUUAUG-----------------CR3Alt3B02\n",
      "\n",
      "UGUUUUUUAUUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUG-UUUAUUUUUUUUUUUUAUUUUAUC--AUU-CR3Conv001\n",
      "---------UUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUGUUUUA-----------UA--UUA-C--A---CR3AltB01\n",
      "---------UUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUGUUUUA-----------UA--UUA-C--A---CR3AltG01\n",
      "---------------------------------AUGGGUUUAUUGUUGUGUUUUA-----------UA--UUA-C--A---CR3AltH01\n",
      "---------UUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUGUUUUA-----------UA--UUA-C--A---CR3AltA02\n",
      "---------UUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUGUUUUA-----------UA--UUA-C--A---CR3AltB02\n",
      "---------------------------------------------------------------------------------CR3AltD01\n",
      "-----------------UUG-UUUUUA-UUAAUAUGGGUUUAUUGUUGUG-UUUA-----------UA--UUA-CUUA---CR3AltC01\n",
      "---------UUGGAUUUUUGUUUUUUAUUUAAUAUGGGUUUAUUGUUGUG-UUUA-----------UA--UUA-CUUA---CR3AltE01\n",
      "---------------------------------------------------------------------------------CR3Alt1D01\n",
      "---------------------------------------------------------------------------------CR3Alt1G01\n",
      "---------------------------------------------------------------------------------CR3Alt1H01\n",
      "---------------------------------------------------------------------------------CR3Alt2A01\n",
      "---------------------------------------------------------------------------------CR3Alt2B01\n",
      "---------------------------------------------------------------------------------CR3Alt2C01\n",
      "---------------------------------------------------------------------------------CR3Alt2D01\n",
      "---------------------------------------------------------------------------------CR3Alt2E01\n",
      "---------------------------------------------------------------------------------CR3Alt2G01\n",
      "---------------------------------------------------------------------------------CR3Alt2H01\n",
      "---------------------------------------------------------------------------------CR3Alt2A02\n",
      "---------------------------------------------------------------------------------CR3Alt2H02\n",
      "---------------------------------------------------------------------------------CR3Alt2B03\n",
      "---------------------------------------------------------------------------------CR3Alt2C03\n",
      "---------------------------------------------------------------------------------CR3Alt2E03\n",
      "---------------------------------------------------------------------------------CR3Alt2F03\n",
      "---------------------------------------------------------------------------------CR3Alt2A04\n",
      "---------------------------------------------------------------------------------CR3Alt2C04\n",
      "---------------------------------------------------------------------------------CR3Alt3A01\n",
      "---------------------------------------------------------------------------------CR3Alt3B01\n",
      "---------------------------------------------------------------------------------CR3Alt3G01\n",
      "---------------------------------------------------------------------------------CR3Alt3B02\n",
      "\n",
      "UG-A----UA-UG-UG-UAUCA-----A--AUUGUUAUUUAUUAUUUA-G-------UUC---GUUUA-UAUUGUUAUUU-CR3Conv001\n",
      "-G-AUUUUUA--G-UGUUAUCA---UUAUUAUUG-UA---G--A---AAA----G-UUUC---G-UUAUUA--G--A----CR3AltB01\n",
      "-G-AUUUUUA--G-UGUUAUCA---UUAUUAUUG-UA--UA-UA---A-G------UUUC---G-UUAUUA--G--A----CR3AltG01\n",
      "-G-AUUUUUA--G-UGUUAUCA---UUAUUAUUG-UA--UA-UA---AG-UUUUC--------G-UUAUUA--G--A----CR3AltH01\n",
      "-G-AUUUUUA--G-UGUUAUCAUUUUUAUUAUUG-UA--UA-UA---A-G-----UUUUC---G-UUAUUA--G--A----CR3AltA02\n",
      "-G-AUUUUUA--G-UGUUAUCAUUUUUAUUAUUG-UA--UA-UA---A-G-----UUUUC---G-UUAUUA--G--A----CR3AltB02\n",
      "---AUUUUUA--G-UGUUAUCAUUUUUAUUAUUG-UA--UA-UA---AG-UUUUC--------G-UUAUUA--G--A----CR3AltD01\n",
      "UGUAUUUUUAUUGUUGUUAUCAUUUUUAUUAUUG-UA--UA-UA---A-G--------UCUUUG-UUAUUA--G--A----CR3AltC01\n",
      "UGUAUUUUUA-----------------------------------------------------------------------CR3AltE01\n",
      "---------------------------------------------------------------------------------CR3Alt1D01\n",
      "---------------------------------------------------------------------------------CR3Alt1G01\n",
      "---------------------------------------------------------------------------------CR3Alt1H01\n",
      "---------------------------------------------------------------------------------CR3Alt2A01\n",
      "---------------------------------------------------------------------------------CR3Alt2B01\n",
      "---------------------------------------------------------------------------------CR3Alt2C01\n",
      "---------------------------------------------------------------------------------CR3Alt2D01\n",
      "---------------------------------------------------------------------------------CR3Alt2E01\n",
      "---------------------------------------------------------------------------------CR3Alt2G01\n",
      "---------------------------------------------------------------------------------CR3Alt2H01\n",
      "---------------------------------------------------------------------------------CR3Alt2A02\n",
      "---------------------------------------------------------------------------------CR3Alt2H02\n",
      "---------------------------------------------------------------------------------CR3Alt2B03\n",
      "---------------------------------------------------------------------------------CR3Alt2C03\n",
      "---------------------------------------------------------------------------------CR3Alt2E03\n",
      "---------------------------------------------------------------------------------CR3Alt2F03\n",
      "---------------------------------------------------------------------------------CR3Alt2A04\n",
      "---------------------------------------------------------------------------------CR3Alt2C04\n",
      "---------------------------------------------------------------------------------CR3Alt3A01\n",
      "---------------------------------------------------------------------------------CR3Alt3B01\n",
      "---------------------------------------------------------------------------------CR3Alt3G01\n",
      "---------------------------------------------------------------------------------CR3Alt3B02\n",
      "\n",
      "UUAUAAUUUAUUUAAGUAUGCAAAUAAUUUUUGU-CR3Conv001\n",
      "UUA-AA---A---AG-UAUGCAAAUAA-UUUU---CR3AltB01\n",
      "UUA-AA---A---AAGUAUGCAAAUAA-UUUUG--CR3AltG01\n",
      "UUA-AA---A---AAGUAUGCAAAUAAUUUUUG--CR3AltH01\n",
      "UUA-AA---A---AAGUAUGCAAAUAA-UUUUG--CR3AltA02\n",
      "UUA-AA---A---AAGUAUGCAAAUAA--UUU---CR3AltB02\n",
      "UUA-AA---A---AAGUAUGCAAAUAAUUUUUG--CR3AltD01\n",
      "UUA-AA---A---AAGUAUGCAAAUAAUUUUUG--CR3AltC01\n",
      "-----------------------------------CR3AltE01\n",
      "-----------------------------------CR3Alt1D01\n",
      "-----------------------------------CR3Alt1G01\n",
      "-----------------------------------CR3Alt1H01\n",
      "-----------------------------------CR3Alt2A01\n",
      "-----------------------------------CR3Alt2B01\n",
      "-----------------------------------CR3Alt2C01\n",
      "-----------------------------------CR3Alt2D01\n",
      "-----------------------------------CR3Alt2E01\n",
      "-----------------------------------CR3Alt2G01\n",
      "-----------------------------------CR3Alt2H01\n",
      "-----------------------------------CR3Alt2A02\n",
      "-----------------------------------CR3Alt2H02\n",
      "-----------------------------------CR3Alt2B03\n",
      "-----------------------------------CR3Alt2C03\n",
      "-----------------------------------CR3Alt2E03\n",
      "-----------------------------------CR3Alt2F03\n",
      "-----------------------------------CR3Alt2A04\n",
      "-----------------------------------CR3Alt2C04\n",
      "-----------------------------------CR3Alt3A01\n",
      "-----------------------------------CR3Alt3B01\n",
      "-----------------------------------CR3Alt3G01\n",
      "-----------------------------------CR3Alt3B02\n",
      "\n",
      "36.6139998436\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start=time.time()\n",
    "print multipleEdSeqAlign(\"CR3allMarkandJames.fa\",\"CR3alternativealignment\",80)\n",
    "end=time.time()\n",
    "print (end-start)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Conclusion\n",
    "### I was able to create a tool that aligned sequences with different numbers of uridines at different positions. It takes 30-45 seconds to run on a group of 20-30 sequences, however this time seems to increase exponentially.  This does not seem like a good tool to use on transcriptomics data obtained from deep sequencing, however, if converted to a more powerful language, this could be possible.  \n",
    "### The score numbers for this tool were difficult to identify.  It seems that giving the gap penalty a greater weight than the mismatch penalty achieves the desired results.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
